### ISSUE TITLE
Builders with > 32 fields do not correctly check if all fields have been initialized

### ISSUE THREAD

**Issue by narbsy** [immutables/immutables#304]
Hey all,

I ran into an interesting side-case the other day.

I have a large-ish interface (30-40 fields) for passing around data. When switching it over to Immutables, I started to rely on the missing field behavior in builders, but noticed some tests failing with NPEs due to a field not being set.

I've reproduced with this example:

```
public class RequiredFieldsTest {
    // fails -- note field #33 is not set.
    @Test
    public void shouldNotAllowPartialModelToBeBuilt() {
        try {
            ImmutableHugeInterface.builder()
                    .number0(0)
                    .number1(1)
                    .number2(2)
                    .number3(3)
                    .number4(4)
                    .number5(5)
                    .number6(6)
                    .number7(7)
                    .number8(8)
                    .number9(9)
                    .number10(10)
                    .number11(11)
                    .number12(12)
                    .number13(13)
                    .number14(14)
                    .number15(15)
                    .number16(16)
                    .number17(17)
                    .number18(18)
                    .number19(19)
                    .number20(20)
                    .number21(21)
                    .number22(22)
                    .number23(23)
                    .number24(24)
                    .number25(25)
                    .number26(26)
                    .number27(27)
                    .number28(28)
                    .number29(29)
                    .number30(30)
                    .number31(31)
                    .build();
            check(false);
        } catch (Exception ex) {
            check(ex.getMessage()).contains("number");
        }
    }

   // Passes
    @Test
    public void shouldNotAllowPartialModelBuiltWithNoFieldsSet() {
        try {
            ImmutableHugeInterface.builder().build();
            check(false);
        } catch (Exception ex) {
            check(ex.getMessage()).contains("number");
        }
    }

    @Value.Immutable
    public interface HugeInterface {
        int getNumber0();
        int getNumber1();
        int getNumber2();
        int getNumber3();
        int getNumber4();
        int getNumber5();
        int getNumber6();
        int getNumber7();
        int getNumber8();
        int getNumber9();
        int getNumber10();
        int getNumber11();
        int getNumber12();
        int getNumber13();
        int getNumber14();
        int getNumber15();
        int getNumber16();
        int getNumber17();
        int getNumber18();
        int getNumber19();
        int getNumber20();
        int getNumber21();
        int getNumber22();
        int getNumber23();
        int getNumber24();
        int getNumber25();
        int getNumber26();
        int getNumber27();
        int getNumber28();
        int getNumber29();
        int getNumber30();
        int getNumber31();
        int getNumber32();
    }
}
```

Looking at the generated code, it seems that the `initBits` roll over around field 32:

```
    private static final long INIT_BIT_NUMBER31 = 0xffffffff80000000L;
    private static final long INIT_BIT_NUMBER32 = 0x1L;
    private long initBits = 0xffffffff;
```

Is this a known limitation of Immutables?

Thanks!

**Comment by elucash** [comment]
Thank you @narbsy for discovering this! Immutables value objects are designed to work with large number of attributes. It is correct that `INIT_BIT_NUMBER32 = 0x1L`, the incorrect part is that it should work against additional `long` fields like `initBits1`, `initBits2`.. etc. I thought we had some tests for this, but those tests don't covered the case after some of the recent changes, apparently.
This definitely should be fixed!

**Comment by narbsy** [comment]
Thanks for the quick response @elucash  :)

I think I'm missing something about the extra `initBits` since a `long` should have 64 bits of information available to use but we're only using 32 bits of it with the current rollover behavior.

Either way, glad it'll be fixed!

**Comment by elucash** [comment]
Indeed, it should not rollover at 32 :) but at 64

**Comment by elucash** [comment]
Found: was `this.mask = 1 << bit;` hence the overflow, should be `this.mask = 1L << bit;`

Thank you again! the fix will be released in 1-2 days, in a meantime you can use snapshot in oss.sonatype repo or build from the source.

**Comment by narbsy** [comment]
Awesome, thanks for the quick turnaround @elucash!

**Comment by elucash** [comment]
Released in v 2.1.16 Thank you!

**Comment by mattomatic** [comment]
We are seeing a possibly related issue here on v2.1.18 -- 

```
import org.immutables.value.Value.Immutable;

public class HugeInterfaceTest {
    public static void main(String[] args) {
        // fails with:
        // java.lang.IllegalStateException: Cannot build HugeInterface, some of required attributes are not set []
        ImmutableHugeInterface.builder()
                .number0(0)
                .number1(1)
                .number2(2)
                .number3(3)
                .number4(4)
                .number5(5)
                .number6(6)
                .number7(7)
                .number8(8)
                .number9(9)
                .number10(10)
                .number11(11)
                .number12(12)
                .number13(13)
                .number14(14)
                .number15(15)
                .number16(16)
                .number17(17)
                .number18(18)
                .number19(19)
                .number20(20)
                .number21(21)
                .number22(22)
                .number23(23)
                .number24(24)
                .number25(25)
                .number26(26)
                .number27(27)
                .number28(28)
                .number29(29)
                .number30(30)
                .number31(31)
                .build();
    }

    @Immutable
    interface HugeInterface {
        int getNumber0();
        int getNumber1();
        int getNumber2();
        int getNumber3();
        int getNumber4();
        int getNumber5();
        int getNumber6();
        int getNumber7();
        int getNumber8();
        int getNumber9();
        int getNumber10();
        int getNumber11();
        int getNumber12();
        int getNumber13();
        int getNumber14();
        int getNumber15();
        int getNumber16();
        int getNumber17();
        int getNumber18();
        int getNumber19();
        int getNumber20();
        int getNumber21();
        int getNumber22();
        int getNumber23();
        int getNumber24();
        int getNumber25();
        int getNumber26();
        int getNumber27();
        int getNumber28();
        int getNumber29();
        int getNumber30();
        int getNumber31();
    }
}
```

**Comment by elucash** [comment]
Yep, the problem was not fully fixed, thank you @mattomatic for revealing this! Currently, the fix applied to the master branch and will be released in the next 2.2 beta or 2.2 final. Do you need this backported to 2.1.x or it's possible to use latest versions?

**Comment by mattomatic** [comment]
Thank you @elucash for the quick turnaround -- it should be possible for us to use the newer versions of immutables which contain the fix.

**Comment by elucash** [comment]
released in 2.2 Thank you!
